### ISSUE TITLE
Enum names break schema in some languages

### ISSUE THREAD

**Issue by Dev-Ditrict-Web** [AxeWP/wp-graphql-gravity-forms#372]
### Description

I'm getting {
  "errors": [
    "Name \"__LE_DU_PRINCE___DOUARD\" must not begin with \"__\", which is reserved by GraphQL introspection."
  ]
}

in the GraphiQL ui. 

### Steps to reproduce

Install WordPress
Install Gravitiforns
Install WP GraphQL
Install WPGraphQL-for-gravity-forms

### Plugin Version

0.12.2

### Gravity Forms Version

2.7.10

### WordPress Version

6.2.2

### WPGraphQL Version

1.14.7

### Additional enviornmental details

php 8.0

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have disabled ALL plugins except for Gravity Forms, WPGraphQL, and WPGraphQL for Gravity Forms

- [X] Yes
- [ ] My issue is with a specific 3rd-party plugin.

**Comment by justlevine** [comment]
Hey @Dev-Ditrict-Web ,

Where exactly are you seeing this error? I.e. what's the step you're taking _after_ you install the plugin?

If it's happening when you are trying to query a specific form, please also include how the form is set up and the GraphQL query you are using.

**Comment by Dev-Ditrict-Web** [comment]
@justlevine 

I just installed the plugins and just went in Graphiql. I did not do any query.

**Comment by justlevine** [comment]
I've been trying to replicate this on a vanilla install with the specs you said (WP 6.2.2, GF 2.7.10, PHP 8.0.26), but everything is working fine.

<img width="1883" alt="image" src="https://github.com/AxeWP/wp-graphql-gravity-forms/assets/29322304/b9e0dd3d-1603-48f4-90f1-c626dafc64ea">

Are you sure you disabled all other extensions besides for WPGraphQL, Gravity Forms core, and this plugin? Or custom code perhaps? 

If so, can you head over to `Site Health` and copy over your `Site info to clipboard` and share it?

<img width="864" alt="image" src="https://github.com/AxeWP/wp-graphql-gravity-forms/assets/29322304/e5e4412c-3ad3-4619-9bd1-bddd0e7510a6">

**Comment by Dev-Ditrict-Web** [comment]
This is what I get when the plugin is not activated
![Screenshot 2023-07-14 at 11 21 49 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/e9c44e4f-78f9-4b01-a4cd-e57daa6c1a92)
![Screenshot 2023-07-14 at 11 21 57 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/bc587e68-74d3-4730-b9b8-28a3258e1a23)

This is what I get when I activate it. 
![Screenshot 2023-07-14 at 11 23 02 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/5d89b71a-04ad-46fa-8b6b-cd616bb9bf9f)
![Screenshot 2023-07-14 at 11 22 54 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/6928bd7a-e8bf-4c5b-af0b-fbde210763e7)

**Comment by Dev-Ditrict-Web** [comment]
I found it.

**Comment by Dev-Ditrict-Web** [comment]
I have the WordPress language setup in french. 

When I put it in French it breaks.

**Comment by Dev-Ditrict-Web** [comment]
![Screenshot 2023-07-14 at 11 26 39 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/c67a2e8c-c8d1-4739-a153-191821b8174a)
![Screenshot 2023-07-14 at 11 26 47 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/2679eb1b-3420-4f52-9757-6c2e25137cf6)
![Screenshot 2023-07-14 at 11 27 00 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/199cfc72-66e6-457e-98e1-ab1a81b98256)
![Screenshot 2023-07-14 at 11 27 16 AM](https://github.com/AxeWP/wp-graphql-gravity-forms/assets/65787883/b3a1d683-c274-4889-9f07-c4df84fda56f)

**Comment by Dev-Ditrict-Web** [comment]
I tried, 

Deutsh and it's good
French and it breaks

**Comment by Dev-Ditrict-Web** [comment]
@justlevine Have you had the chance to look if you change the language setting of WordPress, the plugin breaks ?

**Comment by justlevine** [comment]
Hey @Dev-Ditrict-Web 

I've been able to replicate and determine the source of the issue.

Tl;Dr because GraphQL only supports ASCII characters, accentuated characters get stripped out of enum type names, such as in the `AddressFieldCountryEnum` [here](https://github.com/AxeWP/wp-graphql-gravity-forms/blob/ddb40fd6731e54865397f76b3335a34ec7162321/src/Type/Enum/AddressFieldCountryEnum.php#L47).

As far as an actual fix goes, I don't know when I'll have the free time to address this so I can't promise an ETA. Feel free to submit a Pull Request (or alternatively reach out to me on Slack or via https://axepress.dev/contact if you want to sponsor the work on this).

In the interim, you can use the [`graphql_enum_values` filter](https://www.wpgraphql.com/filters/graphql_enum_values) to remove/replace the ones that are giving you trouble.

**Comment by Dev-Ditrict-Web** [comment]
@justlevine 

How can I find out what Enum breaks the plugin. 

I have look in the graphql_enum_values filter, how do I know which one breaks or not the plugin ?

**Comment by justlevine** [comment]
It's written in the `error` response you screenshotted, and that's will you'll see future ones after you fix that.

Essentially any Enum name that has two non-ascii letters at the front will throw an error, as they get replaced with `_`s

**Comment by Dev-Ditrict-Web** [comment]
@justlevine 

What part does GraphQL does only wanna see ASCII characters ?

In the $values key name, $values [value or description] ?

`
foreach ( $countries as $code => $name ) {
    $values[ WPEnumType::get_safe_name( $code ) ] = [
	'value'       => $name,
	// translators: Country.
	'description' => sprintf( __( '%s .', 'wp-graphql-gravity-forms' ), $name ),
   ];
}
`

**Comment by justlevine** [comment]
The array _key_ ( `WPEnumType::get_safe_name()` is what's responsible for making the string compatible with GraphQL (aka stripping out the non-ASCII chars).)

**Comment by Dev-Ditrict-Web** [comment]
@justlevine 

I was looking in the Country and could'nt find the problem, as soon as I got in the province I saw the error. 

So the solution is to change the get_safe_name function in your plugin or in the WPGraphQL ?

**Comment by justlevine** [comment]
If you're looking to contribute a PR, the solution would be to disable the translations when fetching the province names for conversion. 

If you're looking to solve locally the solution is to check inside `graphql_enum_values`, and either replace the generated name or just str_replace any leading `__`.

**Comment by Dev-Ditrict-Web** [comment]
I was able to do a solution but that would work only for french. 

but it works perfectly. 

I'm mapping all non-ascii caracters to a asci caracters, then replace them and then send the new value to the WPGraphql get_safe_name function. 

Only need for french to do this fix in the AddressFieldProvinceEnum. 

`
foreach ( $provinces as $province ) {

			$accentedCharacters = array(
        'À' => 'A', 'Á' => 'A', 'Â' => 'A', 'Ã' => 'A', 'Ä' => 'A', 'Å' => 'A', 'Æ' => 'AE', 'Ç' => 'C',
        'È' => 'E', 'É' => 'E', 'Ê' => 'E', 'Ë' => 'E', 'Ì' => 'I', 'Í' => 'I', 'Î' => 'I', 'Ï' => 'I',
        'Ð' => 'D', 'Ñ' => 'N', 'Ò' => 'O', 'Ó' => 'O', 'Ô' => 'O', 'Õ' => 'O', 'Ö' => 'O', 'Ø' => 'O',
        'Ù' => 'U', 'Ú' => 'U', 'Û' => 'U', 'Ü' => 'U', 'Ý' => 'Y', 'Þ' => 'TH', 'ß' => 'ss',
        'à' => 'a', 'á' => 'a', 'â' => 'a', 'ã' => 'a', 'ä' => 'a', 'å' => 'a', 'æ' => 'ae', 'ç' => 'c',
        'è' => 'e', 'é' => 'e', 'ê' => 'e', 'ë' => 'e', 'ì' => 'i', 'í' => 'i', 'î' => 'i', 'ï' => 'i',
        'ð' => 'd', 'ñ' => 'n', 'ò' => 'o', 'ó' => 'o', 'ô' => 'o', 'õ' => 'o', 'ö' => 'o', 'ø' => 'o',
        'ù' => 'u', 'ú' => 'u', 'û' => 'u', 'ü' => 'u', 'ý' => 'y', 'þ' => 'th', 'ÿ' => 'y',
    	);

			// Replace accented characters with their ASCII equivalents.
			$asciiString = strtr($province, $accentedCharacters);

			// Remove any remaining non-ASCII characters.
			$asciiString = preg_replace('/[^\x20-\x7E]/u', '', $asciiString);

			$values[ WPEnumType::get_safe_name( $asciiString ) ] = [
				'value'       => $province,
				// translators: Province.
				'description' => sprintf( __( '%s .', 'wp-graphql-gravity-forms' ), $province ),
			];
		}
`

**Comment by Dev-Ditrict-Web** [comment]
@justlevine 

Instead what I would suggest is to do a util function in the plugin, that would implement the caracter mapping, and return the WPEnumType::get_safe_name.

Then in all Enums files just call this util function.

**Comment by justlevine** [comment]
Sadly a utility function that safely transliterates all the different languages that Gravity Forms supports would be untenable to support.

WPGraphQL core will hopefully address this upstream, but my assumption is it will require the `iconv` PHP extension and not a manual string replace.

**Comment by justlevine** [comment]
👀 https://github.com/wp-graphql/wp-graphql/issues/2175

**Comment by Dev-Ditrict-Web** [comment]
I do agree that the best fix would be to change WPGraphQL plugin. But I dought it will arrive soon. 

I did try with iconv but it did not work properly for me. That's why a went for a manual solution. 

I know it's not viable for all languages and not viable in the long run. 

So for now, since we use the plugin in production and if we update it breaks the hole Graphql api. We are stuck to manually do this.

**Comment by Dev-Ditrict-Web** [comment]
@justlevine Thanks for your fast reply and support

**Comment by Dev-Ditrict-Web** [comment]
For other people and documentation. 

Using the graphql_enum_values filter, the key that comes in is already written with the __.

In order to use the filter you have to find the specific key that has the issue and replace it manually. 

Which is in my mind the worst solution but in the same time it's the only solution for maintainability. 

The other solution is to make a duplicate the Enum file that has the issue and put above that modifies 'WPEnumType::get_safe_name( $asciiString )'.

**Comment by Dev-Ditrict-Web** [comment]
Here is my filter for french. 

`
add_filter('graphql_enum_values', function($values){
    foreach ($values as $key => $value) {
        if (str_contains($key, '__LE_DU_PRINCE___DOUARD')) {
            unset($values[$key]);
			      $key = str_replace('__LE_DU_PRINCE___DOUARD', 'ILE_DU_PRINCE_EDOUARD', $key);
            $values[$key] = $value;
        }
    }

    return $values;
});
`

**Comment by justlevine** [comment]
> Here is my filter for french.
> 
> ` add_filter('graphql_enum_values', function($values){ foreach ($values as $key => $value) { if (str_contains($key, '__LE_DU_PRINCE___DOUARD')) { unset($values[$key]); $key = str_replace('__LE_DU_PRINCE___DOUARD', 'ILE_DU_PRINCE_EDOUARD', $key); $values[$key] = $value; } }
> 
> ```
> return $values;
> ```
> 
> }); `

For a more generic approach that doesn't require knowing the failing enum name in advance, you can simply strip out any bad names generically: E.g.

```php
add_filter( 'graphql_enum_values', function( array $values ) {
  foreach ( $values as $name => $config ) {
    if ( str_starts_with( '__', $name ) {
      unset( $values[ $name ] );
      $values[ ltrim( $name, '__' ) ] = $config;
    }
  }
    
  return $values;
}
```

**Comment by justlevine** [comment]
Hey @Dev-Ditrict-Web 

WPGraphQL [v1.17.0](https://github.com/wp-graphql/wp-graphql/releases/tag/v1.17.0) added support for providing transliteration maps that work schema-wide, with the `graphql_pre_format_name` filter.

Using your above snippet as inspiration, you can now do something like this to transliterate _all_ generated GraphQL names.

```php
add_filter( 'graphql_pre_format_name',
  function ( $formatted_name, string $original_name, string $replacement, string $regex ): string {
    // Your character map
    $char_map = [
      'À' => 'A', 'Á' => 'A', 'Â' => 'A', 'Ã' => 'A', 'Ä' => 'A', 'Å' => 'A', 'Æ' => 'AE', 'Ç' => 'C',
      'È' => 'E', 'É' => 'E', 'Ê' => 'E', 'Ë' => 'E', 'Ì' => 'I', 'Í' => 'I', 'Î' => 'I', 'Ï' => 'I',
      'Ð' => 'D', 'Ñ' => 'N', 'Ò' => 'O', 'Ó' => 'O', 'Ô' => 'O', 'Õ' => 'O', 'Ö' => 'O', 'Ø' => 'O',
      'Ù' => 'U', 'Ú' => 'U', 'Û' => 'U', 'Ü' => 'U', 'Ý' => 'Y', 'Þ' => 'TH', 'ß' => 'ss',
      'à' => 'a', 'á' => 'a', 'â' => 'a', 'ã' => 'a', 'ä' => 'a', 'å' => 'a', 'æ' => 'ae', 'ç' => 'c',
      'è' => 'e', 'é' => 'e', 'ê' => 'e', 'ë' => 'e', 'ì' => 'i', 'í' => 'i', 'î' => 'i', 'ï' => 'i',
      'ð' => 'd', 'ñ' => 'n', 'ò' => 'o', 'ó' => 'o', 'ô' => 'o', 'õ' => 'o', 'ö' => 'o', 'ø' => 'o',
      'ù' => 'u', 'ú' => 'u', 'û' => 'u', 'ü' => 'u', 'ý' => 'y', 'þ' => 'th', 'ÿ' => 'y',
    ];
   
     // First transliterate the values.
    $transliterated = str_replace( array_keys( $char_map ), array_values( $char_map ), $name ); 
    
    // Then run the standard regex.
    return preg_replace( $regex, $replacement, $transliterated );
  },
  10,
  4
);
```

(Leaving this open until I can add a docs note).

**Comment by Dev-Ditrict-Web** [comment]
@justlevine I'm glad to ear that.  !

Glad I could help.
